# Installlations

In [ ]:
!pip uninstall torch-scatter torch-sparse torch-geometric torch-cluster  
!pip install torch-scatter -f https://data.pyg.org/whl/torch-${TORCH}+${CUDA}.html
!pip install torch-sparse -f https://data.pyg.org/whl/torch-${TORCH}+${CUDA}.html
!pip install torch-geometric
#!pip install -U sentence-transformers

Found existing installation: torch-scatter 2.0.9
Uninstalling torch-scatter-2.0.9:
  Would remove:
    /usr/local/lib/python3.7/dist-packages/torch_scatter-2.0.9.dist-info/*
    /usr/local/lib/python3.7/dist-packages/torch_scatter/*
Proceed (y/n)?   ERROR: Operation cancelled by user
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in links: https://data.pyg.org/whl/torch-+.html
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in links: https://data.pyg.org/whl/torch-+.html
ERROR: Operation cancelled by user
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 3021, in _dep_map
    return self.__dep_map
  File "/usr/local/lib/python3.7/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 2815, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During 

In [ ]:
!pip install optuna

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# Model definitions

In [ ]:
import torch
from torch_geometric.datasets import UPFD
import numpy as np
from torch_geometric.loader import DataLoader

In [ ]:
import torch.nn.functional as F
from torch.nn import ReLU, LeakyReLU, Softmax, Linear, SELU, GELU, ELU, Dropout
from torch_geometric.nn import SAGEConv, global_max_pool, TopKPooling

In [ ]:
from torch.autograd import Variable
from sklearn.metrics import accuracy_score, f1_score 

In [ ]:
test_data_pol = UPFD(root=".", name="politifact", feature="bert",split="test")
train_data_pol = UPFD(root=".", name="politifact", feature="bert", split="train")
val_data_pol = UPFD(root=".", name="politifact", feature="bert", split="val")

In [ ]:
# combined = test_data_pol+train_data_pol+val_data_pol

# train_data_pol,val_data_pol,test_data_pol = torch.utils.data.random_split(combined, [220,31,63],generator=torch.Generator().manual_seed(42))

In [ ]:
train_data_pol

UPFD(62, name=politifact, feature=bert)

In [ ]:
test_loader = DataLoader(test_data_pol, batch_size=32, shuffle=True)
train_loader = DataLoader(train_data_pol, batch_size=32, shuffle=True)
val_loader = DataLoader(val_data_pol, batch_size=32, shuffle=True)

In [ ]:
class Net(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels):
        super(Net, self).__init__()
        self.conv1 = SAGEConv(in_channels, hidden_channels[0])
        self.leaky1 = LeakyReLU()
        self.conv2 = SAGEConv(hidden_channels[0], hidden_channels[1])
        self.leaky2 = LeakyReLU()
        self.conv3 = SAGEConv(hidden_channels[1], hidden_channels[2])
        self.leaky3 = LeakyReLU() 
        
        self.full1 = Linear(hidden_channels[2],hidden_channels[3])
        self.ge1 = GELU()
        self.full2 = Linear(hidden_channels[3],hidden_channels[4])
        self.ge2 = GELU()
        self.full3 = Linear(hidden_channels[4],out_channels)


    def forward(self, x, edge_index, batch):
        h = self.leaky1(self.conv1(x, edge_index))
        h = self.leaky2(self.conv2(h, edge_index))
        h = self.leaky3(self.conv3(h, edge_index))

        h = global_max_pool(h,batch)

        h = self.ge1(self.full1(h))
        h = self.ge2(self.full2(h))
        h = self.full3(h)

        return torch.sigmoid(h)

In [ ]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Hyper paarmeter tuning

In [ ]:
def objective(trial):
  
  model = Net(768,[trial.suggest_int(name="layer_size1",low=256,high=512,step=128),
                   trial.suggest_int(name="layer_size2",low=256,high=512,step=128),
                   trial.suggest_int(name="layer_size3",low=256,high=512,step=128),
                   trial.suggest_int(name="layer_size4",low=64,high=256,step=64),
                   trial.suggest_int(name="layer_size5",low=64,high=256,step=64),
  
  ],1).to(device)
  optimizer = torch.optim.Adam(model.parameters(),lr=trial.suggest_loguniform('learning_rate', 1e-6, 1e-3),betas=(trial.suggest_loguniform('b1', 1-1e-1,1-1e-3),0.99))
  lossff = torch.nn.BCELoss()

  total_loss = 0
  weighted_loss = 0
  exp_param = 0.8

  wloss = []
  best_val = 1

  for i in range(1600):
    print("Epoch:", i)
    model.train()
    for data in train_loader:
          data = data.to(device)
          optimizer.zero_grad()
          out = model(data.x, data.edge_index, data.batch)
          loss = lossff(torch.reshape(out,(-1,)), data.y.float())
          loss.backward()
          optimizer.step()
          total_loss += float(loss) * data.num_graphs
    print("Train: ",total_loss / len(train_loader.dataset))

    model.eval()
    total_loss = 0
    for data in val_loader:
          data = data.to(device)
          out = model(data.x, data.edge_index, data.batch)
          loss = lossff(torch.reshape(out,(-1,)), data.y.float())
          total_loss += float(loss) * data.num_graphs
          
    print("Test", total_loss / len(val_loader.dataset))

    weighted_loss = exp_param*(weighted_loss) + (1-exp_param)*(total_loss/ len(val_loader.dataset))
    print(weighted_loss/(1-exp_param**(i+1)))
    wloss.append(weighted_loss/(1-exp_param**(i+1)))

    if(best_val>weighted_loss/(1-exp_param**(i+1))):
      best_val = weighted_loss/(1-exp_param**(i+1))

    if(i-10>=0 and wloss[i-10]-weighted_loss<-0.01):
      break

  return best_val

In [ ]:
import optuna
study = optuna.create_study(direction="minimize", sampler=optuna.samplers.TPESampler())
study.optimize(objective, n_trials=20)

[I 2022-10-16 08:16:04,478] A new study created in memory with name: no-name-1e8ef305-5212-47c4-a7fd-4ef3306f7435
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :func:`~optuna.trial.Trial.suggest_float` instead.
  # Remove the CWD from sys.path while we load stuff.


Epoch: 0
Train:  0.6879171094586772
Test 0.699082612991333
0.699082612991333
Epoch: 1
Train:  1.0370765482225726
Test 0.6996992826461792
0.6994252072440253
Epoch: 2
Train:  1.03695761388348
Test 0.700280487537384
0.6997757319544183
Epoch: 3
Train:  1.0366437098672312
Test 0.7006199955940247
0.7000617291223066
Epoch: 4
Train:  1.0365300899551761
Test 0.7010000348091125
0.700340853850652
Epoch: 5
Train:  1.0366294220570595
Test 0.7014452815055847
0.7006402151501285
Epoch: 6
Train:  1.0363946374385589
Test 0.7018416523933411
0.7009442668649565
Epoch: 7
Train:  1.036256968013702
Test 0.7022421360015869
0.7012561692249233
Epoch: 8
Train:  1.0362733775569546
Test 0.7026827931404114
0.7015857264269187
Epoch: 9
Train:  1.036109584954477
Test 0.7031116485595703
0.701927621564291
Epoch: 10
Train:  1.035969365027643
Test 0.7035372257232666
0.7022797937570996
Epoch: 11
Train:  1.0359365747820946
Test 0.7040051817893982
0.7026503347464571
Epoch: 12
Train:  1.0359328783327533
Test 0.7044985890388489

[I 2022-10-16 08:16:12,161] Trial 0 finished with value: 0.5168222606771309 and parameters: {'layer_size1': 384, 'layer_size2': 256, 'layer_size3': 512, 'layer_size4': 256, 'layer_size5': 128, 'learning_rate': 1.7269151676184985e-05, 'b1': 0.9494564658986889}. Best is trial 0 with value: 0.5168222606771309.


Train:  0.3164188884679348
Test 0.5362651944160461
0.5245158449125862
Epoch: 166
Train:  0.31888307823288825
Test 0.5446001887321472
0.5285327136764985
Epoch: 167
Train:  0.32077544795409324
Test 0.5491675138473511
0.5326596737106689
Epoch: 0
Train:  0.6874574403608998
Test 0.7044631838798523
0.7044631838798522
Epoch: 1
Train:  1.0372989129635595
Test 0.7082279324531555
0.7065547108650208
Epoch: 2
Train:  1.0367923965377193
Test 0.712127685546875
0.7088387168821741
Epoch: 3
Train:  1.0368902241030047
Test 0.7167356610298157
0.7115138334633535
Epoch: 4
Train:  1.0375670627240212
Test 0.7220823764801025
0.7146577360742379
Epoch: 5
Train:  1.0389100197822816
Test 0.7286458015441895
0.7184492795379902
Epoch: 6
Train:  1.0401950324735334
Test 0.7325580716133118
0.7220198384239315
Epoch: 7
Train:  1.0410758074252837
Test 0.7357912063598633
0.7253293571095673
Epoch: 8
Train:  1.0421648871514104
Test 0.7399211525917053
0.7287001342562118
Epoch: 9
Train:  1.0425259172916412
Test 0.7425117492675

[I 2022-10-16 08:16:14,395] Trial 1 finished with value: 0.5954291380651352 and parameters: {'layer_size1': 256, 'layer_size2': 512, 'layer_size3': 384, 'layer_size4': 192, 'layer_size5': 256, 'learning_rate': 9.006247606734518e-05, 'b1': 0.9747253956189821}. Best is trial 0 with value: 0.5168222606771309.


Train:  0.37677071003183243
Test 0.7050283551216125
0.6419269484951498
Epoch: 0
Train:  0.6978581989965131
Test 0.6896348595619202
0.6896348595619202
Epoch: 1
Train:  1.0423623948327956
Test 0.6899744868278503
0.6898235413763258
Epoch: 2
Train:  1.0420034152846183
Test 0.6903107762336731
0.6900232277932714
Epoch: 3
Train:  1.0416451779104048
Test 0.690634548664093
0.6902303148091324
Epoch: 4
Train:  1.0414709496882655
Test 0.6909946799278259
0.6904576961509618
Epoch: 5
Train:  1.0410965786826225
Test 0.6913432478904724
0.6906977299080935
Epoch: 6
Train:  1.0407570014076848
Test 0.6916397213935852
0.6909361228230253
Epoch: 7
Train:  1.0406813650361952
Test 0.6919804811477661
0.6911871017676231
Epoch: 8
Train:  1.040470011772648
Test 0.6923360228538513
0.6914525082383688
Epoch: 9
Train:  1.04014699401394
Test 0.6926824450492859
0.6917280854817481
Epoch: 10
Train:  1.0400564314857605
Test 0.6930604577064514
0.6920196009114234
Epoch: 11
Train:  1.0397635650250219
Test 0.6934406757354736
0.

[I 2022-10-16 08:16:24,256] Trial 2 finished with value: 0.5052036209126595 and parameters: {'layer_size1': 256, 'layer_size2': 384, 'layer_size3': 512, 'layer_size4': 192, 'layer_size5': 192, 'learning_rate': 1.2532950729993314e-05, 'b1': 0.9151613761338718}. Best is trial 2 with value: 0.5052036209126595.


Test 0.5238955020904541
0.5180004152398721
Epoch: 235
Train:  0.31373049667285335
Test 0.5262512564659119
0.51965058348508
Epoch: 236
Train:  0.31355854977042447
Test 0.5279003381729126
0.5213005344226465
Epoch: 0
Train:  0.693409348687818
Test 0.6935471296310425
0.6935471296310425
Epoch: 1
Train:  1.0390670011120458
Test 0.6943220496177673
0.6939776407347785
Epoch: 2
Train:  1.0389376330760218
Test 0.6954788565635681
0.6945928931236267
Epoch: 3
Train:  1.037984614410708
Test 0.6963846683502197
0.6951998630513343
Epoch: 4
Train:  1.0376540576258013
Test 0.6974413394927979
0.6958666516167388
Epoch: 5
Train:  1.0375247886103969
Test 0.6987148523330688
0.6966386723677608
Epoch: 6
Train:  1.0370301950362422
Test 0.7000693678855896
0.6975068898806951
Epoch: 7
Train:  1.0365764027641666
Test 0.701513946056366
0.6984698608324168
Epoch: 8
Train:  1.0366702262432343
Test 0.7032120823860168
0.6995653375375809
Epoch: 9
Train:  1.0362175693435054
Test 0.7050349116325378
0.7007908396431917
Epoch: 1

[I 2022-10-16 08:16:25,115] Trial 3 finished with value: 0.6935471296310425 and parameters: {'layer_size1': 256, 'layer_size2': 512, 'layer_size3': 384, 'layer_size4': 192, 'layer_size5': 256, 'learning_rate': 3.818955426351249e-05, 'b1': 0.9449134398799891}. Best is trial 2 with value: 0.5052036209126595.


Streaming output truncated to the last 5000 lines.
Test 0.703769326210022
0.7036458093416699
Epoch: 347
Train:  1.0349521348553319
Test 0.7038022875785828
0.7036771049890524
Epoch: 348
Train:  1.034947955800641
Test 0.7038379311561584
0.7037092702224736
Epoch: 349
Train:  1.0349329219710441
Test 0.7038714289665222
0.7037417019712834
Epoch: 350
Train:  1.0349239505106402
Test 0.7039053440093994
0.7037744303789066
Epoch: 351
Train:  1.034907535199196
Test 0.7039365768432617
0.7038068596717776
Epoch: 352
Train:  1.0349089810925145
Test 0.7039738893508911
0.7038402656076004
Epoch: 353
Train:  1.0348992962991037
Test 0.7040097713470459
0.7038741667554894
Epoch: 354
Train:  1.0348805900542968
Test 0.7040411829948425
0.70390757000336
Epoch: 355
Train:  1.0348773339102346
Test 0.7040765881538391
0.7039413736334559
Epoch: 356
Train:  1.0348596678626152
Test 0.7041081786155701
0.7039747346298787
Epoch: 357
Train:  1.0348591641072304
Test 0.7041436433792114
0.7040085163797453
Epoch: 358
Train:  1

[I 2022-10-16 08:17:30,640] Trial 4 finished with value: 0.6948789358139038 and parameters: {'layer_size1': 256, 'layer_size2': 256, 'layer_size3': 384, 'layer_size4': 128, 'layer_size5': 128, 'learning_rate': 1.2244078258235934e-06, 'b1': 0.9059061503282786}. Best is trial 2 with value: 0.5052036209126595.


Test 0.7184032201766968
0.7186025116480687
Epoch: 1597
Train:  0.972076544838567
Test 0.7183598279953003
0.7185539749175149
Epoch: 1598
Train:  0.9719325304031372
Test 0.7183486819267273
0.7185129163193574
Epoch: 1599
Train:  0.9717830563745191
Test 0.7182978391647339
0.7184699008884328
Epoch: 0
Train:  0.6911747628642667
Test 0.695466935634613
0.695466935634613
Epoch: 1
Train:  1.0386938750743866
Test 0.6955997943878174
0.69554074605306
Epoch: 2
Train:  1.0385985643632951
Test 0.6957418322563171
0.6956231584314441
Epoch: 3
Train:  1.0385058358792336
Test 0.6958914995193481
0.695714059883986
Epoch: 4
Train:  1.0384155531083383
Test 0.696044385433197
0.6958123242667832
Epoch: 5
Train:  1.0383472606059043
Test 0.69620680809021
0.6959192513158082
Epoch: 6
Train:  1.0382842979123514
Test 0.696378767490387
0.6960355426008181
Epoch: 7
Train:  1.0382085106065195
Test 0.69655442237854
0.6961602391689411
Epoch: 8
Train:  1.038167695845327
Test 0.696729302406311
0.6962916955999109
Epoch: 9
Train

[I 2022-10-16 08:18:12,551] Trial 5 finished with value: 0.48990411727913946 and parameters: {'layer_size1': 256, 'layer_size2': 512, 'layer_size3': 512, 'layer_size4': 128, 'layer_size5': 64, 'learning_rate': 5.478257874192845e-06, 'b1': 0.958444725230597}. Best is trial 5 with value: 0.48990411727913946.


Streaming output truncated to the last 5000 lines.
Train:  1.0183528661727905
Test 0.7223283052444458
0.7223304655549172
Epoch: 321
Train:  1.0181746540531036
Test 0.7223402261734009
0.7223324176786139
Epoch: 322
Train:  1.017995415195342
Test 0.7223573923110962
0.7223374126051103
Epoch: 323
Train:  1.017784099425039
Test 0.7223445177078247
0.7223388336256532
Epoch: 324
Train:  1.017574344911883
Test 0.7223268151283264
0.7223364299261879
Epoch: 325
Train:  1.017368265698033
Test 0.7223024368286133
0.7223296313066729
Epoch: 326
Train:  1.0171648917659637
Test 0.7222942113876343
0.7223225473228652
Epoch: 327
Train:  1.016953545231973
Test 0.7222746014595032
0.7223129581501928
Epoch: 328
Train:  1.016738106166163
Test 0.7222471237182617
0.7222997912638066
Epoch: 329
Train:  1.01652539930036
Test 0.722233235836029
0.7222864801782511
Epoch: 330
Train:  1.0163210582348607
Test 0.7222238183021545
0.7222739478030319
Epoch: 331
Train:  1.016100303780648
Test 0.722195029258728
0.7222581640941712

[I 2022-10-16 08:19:20,845] Trial 6 finished with value: 0.4904637024548766 and parameters: {'layer_size1': 512, 'layer_size2': 256, 'layer_size3': 384, 'layer_size4': 192, 'layer_size5': 128, 'learning_rate': 3.50987342986947e-06, 'b1': 0.9669303847062202}. Best is trial 5 with value: 0.48990411727913946.


Streaming output truncated to the last 5000 lines.
Train:  1.0355654570364183
Test 0.7147889733314514
0.7145629977909171
Epoch: 349
Train:  1.0355687977806214
Test 0.7148469090461731
0.7146197800419684
Epoch: 350
Train:  1.0355506971959145
Test 0.7149003744125366
0.714675898916082
Epoch: 351
Train:  1.0355511884535513
Test 0.7149568200111389
0.7147320831350934
Epoch: 352
Train:  1.035540618242756
Test 0.7150121927261353
0.7147881050533017
Epoch: 353
Train:  1.0355429264806932
Test 0.7150701284408569
0.7148445097308127
Epoch: 354
Train:  1.035526983199581
Test 0.7151249647140503
0.7149006007274602
Epoch: 355
Train:  1.0355288578617958
Test 0.7151828408241272
0.7149570487467936
Epoch: 356
Train:  1.0355318352099387
Test 0.7152420282363892
0.7150140446447127
Epoch: 357
Train:  1.0355120474292385
Test 0.7152968049049377
0.7150705966967577
Epoch: 358
Train:  1.035505634161734
Test 0.7153518199920654
0.7151268413558192
Epoch: 359
Train:  1.0355119512927147
Test 0.7154106497764587
0.715183603

[I 2022-10-16 08:20:29,272] Trial 7 finished with value: 0.6281976435602985 and parameters: {'layer_size1': 384, 'layer_size2': 256, 'layer_size3': 512, 'layer_size4': 128, 'layer_size5': 192, 'learning_rate': 1.5683612126291248e-06, 'b1': 0.9679857583157938}. Best is trial 5 with value: 0.48990411727913946.


Streaming output truncated to the last 5000 lines.
Train:  1.0281391951345629
Test 0.7346819639205933
0.7346087568612865
Epoch: 349
Train:  1.0280144810676575
Test 0.7346898913383484
0.7346249837566988
Epoch: 350
Train:  1.0278944786517852
Test 0.7347084283828735
0.7346416726819338
Epoch: 351
Train:  1.0277976278335816
Test 0.7347443103790283
0.7346622002213526
Epoch: 352
Train:  1.0276674012984
Test 0.7347562313079834
0.7346810064386787
Epoch: 353
Train:  1.027544940671613
Test 0.7347531318664551
0.734695431524234
Epoch: 354
Train:  1.0274251949402593
Test 0.7347759008407593
0.734711525387539
Epoch: 355
Train:  1.0272967584671513
Test 0.7347820401191711
0.7347256283338655
Epoch: 356
Train:  1.0271701764675878
Test 0.734762966632843
0.7347330959936609
Epoch: 357
Train:  1.0270346681917868
Test 0.734769880771637
0.734740452949256
Epoch: 358
Train:  1.0269198542641056
Test 0.7347885370254517
0.7347500697644952
Epoch: 359
Train:  1.026803958800531
Test 0.7348031997680664
0.734760695765209

[I 2022-10-16 08:21:37,246] Trial 8 finished with value: 0.5081802207654396 and parameters: {'layer_size1': 384, 'layer_size2': 256, 'layer_size3': 512, 'layer_size4': 192, 'layer_size5': 128, 'learning_rate': 3.181530760768166e-06, 'b1': 0.9791175824385046}. Best is trial 5 with value: 0.48990411727913946.


Train:  0.3843325168840707
Test 0.7668970227241516
0.7635339932040688
Epoch: 1599
Train:  0.3847327927623184
Test 0.7675942182540894
0.7643460382140729
Epoch: 0
Train:  0.7004362633151393
Test 0.6879337430000305
0.6879337430000305
Epoch: 1
Train:  1.0437591835375755
Test 0.6883358955383301
0.6881571610768638
Epoch: 2
Train:  1.0433000749157322
Test 0.6887269616127014
0.6883906858866332
Epoch: 3
Train:  1.0427323762447602
Test 0.6891233325004578
0.6886388724089314
Epoch: 4
Train:  1.0422470386951201
Test 0.6895303726196289
0.6889040735663261
Epoch: 5
Train:  1.042054199403332
Test 0.6899598240852356
0.689190240655544
Epoch: 6
Train:  1.041421760474482
Test 0.6903814077377319
0.6894916932665996
Epoch: 7
Train:  1.0410426009085871
Test 0.6908157467842102
0.6898098882257133
Epoch: 8
Train:  1.0408368043361171
Test 0.6912841200828552
0.690150443098415
Epoch: 9
Train:  1.0404679573351336
Test 0.6917787194252014
0.6905152715169262
Epoch: 10
Train:  1.0399411253390773
Test 0.6922900080680847
0

[I 2022-10-16 08:21:38,781] Trial 9 finished with value: 0.6879337430000305 and parameters: {'layer_size1': 256, 'layer_size2': 256, 'layer_size3': 512, 'layer_size4': 256, 'layer_size5': 64, 'learning_rate': 1.7950294778412504e-05, 'b1': 0.9876916157544144}. Best is trial 5 with value: 0.48990411727913946.


Train:  1.0356388870746858
Test 0.7141029834747314
0.7093646244453128
Epoch: 36
Train:  1.0354215964194267
Test 0.715388834476471
0.7105697793276399
Epoch: 0
Train:  0.7028722647697695
Test 0.7069706916809082
0.7069706916809082
Epoch: 1
Train:  1.0452177736067003
Test 0.7395938038825989
0.7250946429040698
Epoch: 2
Train:  1.0576597404095434
Test 0.7473073601722717
0.7341982155549721
Epoch: 3
Train:  1.0595572677350813
Test 0.715994656085968
0.7280316981196726
Epoch: 4
Train:  1.0294423574401486
Test 0.711311399936676
0.7230577874274438
Epoch: 5
Train:  1.021332420649067
Test 0.7131586074829102
0.7203745592787172
Epoch: 6
Train:  1.0125263640957494
Test 0.7244599461555481
0.7214084617090386
Epoch: 7
Train:  0.9970934938999915
Test 0.7341088652610779
0.7244606075137354
Epoch: 8
Train:  0.9739578202847512
Test 0.7150238752365112
0.7222806755418264
Epoch: 9
Train:  0.9232139029810505
Test 0.6997758150100708
0.7172382803773523
Epoch: 10
Train:  0.8644496644696882
Test 0.6808726787567139
0.7

[I 2022-10-16 08:21:39,509] Trial 10 finished with value: 0.704185272738625 and parameters: {'layer_size1': 512, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 0.0007298619879847754, 'b1': 0.9313358142210422}. Best is trial 5 with value: 0.48990411727913946.


Streaming output truncated to the last 5000 lines.
Train:  0.9614702828468815
Test 0.7109917402267456
0.7119896893023003
Epoch: 346
Train:  0.9606557488441467
Test 0.7107765078544617
0.7117470530127326
Epoch: 347
Train:  0.9598519465615672
Test 0.7105420827865601
0.7115060589674981
Epoch: 348
Train:  0.9590731301615315
Test 0.7102640271186829
0.711257652597735
Epoch: 349
Train:  0.9582301907001003
Test 0.7099941372871399
0.711004949535616
Epoch: 350
Train:  0.9574190464711958
Test 0.7095962166786194
0.7107232029642166
Epoch: 351
Train:  0.9564914289982088
Test 0.7092248201370239
0.7104235263987781
Epoch: 352
Train:  0.9556009557939344
Test 0.7089070677757263
0.7101202346741677
Epoch: 353
Train:  0.9547254029781588
Test 0.7085942625999451
0.7098150402593232
Epoch: 354
Train:  0.9538698878980452
Test 0.7082469463348389
0.7095014214744264
Epoch: 355
Train:  0.9530002628603289
Test 0.7077785134315491
0.709156839865851
Epoch: 356
Train:  0.9520102498992797
Test 0.7074641585350037
0.70881830

[I 2022-10-16 08:22:54,410] Trial 11 finished with value: 0.46340372127499985 and parameters: {'layer_size1': 512, 'layer_size2': 512, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 4.531486810010387e-06, 'b1': 0.9591472429391925}. Best is trial 11 with value: 0.46340372127499985.


Train:  0.4573364872731843
Test 0.9156977534294128
0.9126845880633453
Epoch: 1596
Train:  0.4578767312603656
Test 0.9165862202644348
0.9134649145035633
Epoch: 1597
Train:  0.45832065413566353
Test 0.9175319671630859
0.9142783250354678
Epoch: 1598
Train:  0.4587932631917107
Test 0.9188184142112732
0.9151863428706289
Epoch: 1599
Train:  0.4594361790612076
Test 0.9199652671813965
0.9161421277327825
Epoch: 0
Train:  0.6918934487527416
Test 0.6945173144340515
0.6945173144340515
Epoch: 1
Train:  1.0389655549680032
Test 0.6946743130683899
0.6946045358975729
Epoch: 2
Train:  1.0388197273977342
Test 0.6948102116584778
0.694688829242206
Epoch: 3
Train:  1.0387186763748046
Test 0.694945752620697
0.6947758629070063
Epoch: 4
Train:  1.0386789335358528
Test 0.6950809359550476
0.6948666152416212
Epoch: 5
Train:  1.0386230205335925
Test 0.6951943039894104
0.6949554371114142
Epoch: 6
Train:  1.0384924921297258
Test 0.6952911615371704
0.6950404000072604
Epoch: 7
Train:  1.038395320215533
Test 0.69538694

[I 2022-10-16 08:23:43,273] Trial 12 finished with value: 0.49886128599702423 and parameters: {'layer_size1': 512, 'layer_size2': 512, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 6.1013691385940126e-06, 'b1': 0.9508180964022637}. Best is trial 11 with value: 0.46340372127499985.


Test 0.8368896842002869
0.831935091326876
Epoch: 1042
Train:  0.4187736939845396
Test 0.8381718397140503
0.8331824410043109
Epoch: 1043
Train:  0.4194109659058027
Test 0.8390942811965942
0.8343648090427676
Epoch: 1044
Train:  0.4198685982996868
Test 0.8398234248161316
0.8354565321974403
Epoch: 1045
Train:  0.4202295336478029
Test 0.8400970697402954
0.8363846397060114
Epoch: 0
Train:  0.6918606604299238
Test 0.6967666745185852
0.6967666745185852
Epoch: 1
Train:  1.03842860940964
Test 0.6987465620040894
0.6978666120105321
Epoch: 2
Train:  1.0369776506577768
Test 0.7002087235450745
0.6988264937869839
Epoch: 3
Train:  1.036668130467015
Test 0.7020255923271179
0.6999101992545089
Epoch: 4
Train:  1.0359314574349312
Test 0.703866183757782
0.7010870152062204
Epoch: 5
Train:  1.0355087220668793
Test 0.7058256268501282
0.7023714424234303
Epoch: 6
Train:  1.0351267024394004
Test 0.7080250382423401
0.7038022167009844
Epoch: 7
Train:  1.035038681760911
Test 0.7106346487998962
0.705444178618267
Epoc

[I 2022-10-16 08:23:44,247] Trial 13 finished with value: 0.6967666745185852 and parameters: {'layer_size1': 512, 'layer_size2': 512, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 8.251405583050624e-05, 'b1': 0.9333946424613304}. Best is trial 11 with value: 0.46340372127499985.


Train:  1.0254000386884135
Test 0.7335594892501831
0.728830953958704
Epoch: 0
Train:  0.6869825740014354
Test 0.7019327282905579
0.7019327282905579
Epoch: 1
Train:  1.037777949725428
Test 0.702113687992096
0.7020332614580791
Epoch: 2
Train:  1.0376828399396711
Test 0.7022709846496582
0.7021306889956115
Epoch: 3
Train:  1.0376405831306212
Test 0.7024465799331665
0.7022376981207994
Epoch: 4
Train:  1.037553979504493
Test 0.7025817632675171
0.7023400497232261
Epoch: 5
Train:  1.0375195703198832
Test 0.702732503414154
0.7024464264934708
Epoch: 6
Train:  1.0374671480348032
Test 0.7028886675834656
0.7025583459154298
Epoch: 7
Train:  1.0373933863255285
Test 0.703031599521637
0.7026720776565374
Epoch: 8
Train:  1.0373561526498487
Test 0.7031781673431396
0.7027889868848665
Epoch: 9
Train:  1.0373188660990806
Test 0.7033330798149109
0.7029108952857476
Epoch: 10
Train:  1.037302698819868
Test 0.7034972310066223
0.7030391821911901
Epoch: 11
Train:  1.0372262356742736
Test 0.703652024269104
0.70317

[I 2022-10-16 08:24:14,954] Trial 14 finished with value: 0.4876034455782996 and parameters: {'layer_size1': 384, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 128, 'layer_size5': 64, 'learning_rate': 6.987237846980255e-06, 'b1': 0.959942400022472}. Best is trial 11 with value: 0.46340372127499985.


Train:  0.303778656803432
Test 0.5922014713287354
0.5880581969024117
Epoch: 709
Train:  0.3041542945129256
Test 0.5929266214370728
0.5890318818093438
Epoch: 0
Train:  0.6882275073759018
Test 0.7075703144073486
0.7075703144073486
Epoch: 1
Train:  1.0336663569173505
Test 0.7283786535263062
0.7191305028067695
Epoch: 2
Train:  1.049343922445851
Test 0.7529381513595581
0.7329860964759451
Epoch: 3
Train:  1.0521191570066637
Test 0.7489702701568604
0.7384007894572849
Epoch: 4
Train:  1.0447523343947627
Test 0.7320171594619751
0.7365018038565859
Epoch: 5
Train:  1.030330244571932
Test 0.7201286554336548
0.7320637703045294
Epoch: 6
Train:  1.014624128418584
Test 0.7180479764938354
0.7285167467983974
Epoch: 7
Train:  0.9978680860611701
Test 0.7183259129524231
0.7260676977865368
Epoch: 8
Train:  0.9760895938642563
Test 0.7232518196105957
0.7254172159582507
Epoch: 9
Train:  0.9514498037676657
Test 0.7230282425880432
0.7248819472436104
Epoch: 10
Train:  0.9133175275018138
Test 0.6973909139633179
0.

[I 2022-10-16 08:24:15,945] Trial 15 finished with value: 0.6682903692304043 and parameters: {'layer_size1': 384, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 0.0005309941801087301, 'b1': 0.9912420385977869}. Best is trial 11 with value: 0.46340372127499985.


Train:  0.49044660406727947
Test 0.728853166103363
0.6806251319358927
Epoch: 18
Train:  0.47241191998604803
Test 0.6793131232261658
0.6803588932907049
Epoch: 19
Train:  0.40635827436081823
Test 0.7234585285186768
0.6890793604778015
Epoch: 20
Train:  0.3970282343967307
Test 0.9062542915344238
0.7329186931643412
Epoch: 21
Train:  0.4720116129685794
Test 1.0236830711364746
0.7915038509234481
Epoch: 0
Train:  0.6957967319796162
Test 0.6910147070884705
0.6910147070884705
Epoch: 1
Train:  1.041092440966637
Test 0.6911373734474182
0.6910828550656637
Epoch: 2
Train:  1.0409840528042085
Test 0.6912848353385925
0.6911656338660443
Epoch: 3
Train:  1.0408379925835518
Test 0.6914350390434265
0.6912568957825016
Epoch: 4
Train:  1.0407484548707162
Test 0.6915946006774902
0.6913573553538332
Epoch: 5
Train:  1.0406266758518834
Test 0.6917515397071838
0.6914642012298174
Epoch: 6
Train:  1.040532241905889
Test 0.6918973922729492
0.6915738303263486
Epoch: 7
Train:  1.0403561438283613
Test 0.69203162193298

[I 2022-10-16 08:24:33,592] Trial 16 finished with value: 0.4985875173504829 and parameters: {'layer_size1': 384, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 128, 'layer_size5': 192, 'learning_rate': 9.339684379511036e-06, 'b1': 0.9372308065815763}. Best is trial 11 with value: 0.46340372127499985.


Streaming output truncated to the last 5000 lines.
Train:  1.0351628786133182
Test 0.7177742719650269
0.7175125479606698
Epoch: 346
Train:  1.0351806456042874
Test 0.7178395390510559
0.7175779461787469
Epoch: 347
Train:  1.0351697754475377
Test 0.7179051041603088
0.7176433777750593
Epoch: 348
Train:  1.0351609116600407
Test 0.71797114610672
0.7177089314413914
Epoch: 349
Train:  1.035153074610618
Test 0.7180372476577759
0.7177745946846682
Epoch: 350
Train:  1.0351723144131322
Test 0.7181029319763184
0.7178402621429982
Epoch: 351
Train:  1.035140210582364
Test 0.7181671261787415
0.7179056349501469
Epoch: 352
Train:  1.0351653531674416
Test 0.718230664730072
0.7179706409061319
Epoch: 353
Train:  1.0351498771098353
Test 0.7182942032814026
0.7180353533811861
Epoch: 354
Train:  1.0351497213686667
Test 0.7183581590652466
0.7180999145179982
Epoch: 355
Train:  1.0351304738752303
Test 0.7184219360351562
0.7181643188214298
Epoch: 356
Train:  1.0351474804262961
Test 0.7184858918190002
0.7182286334

[I 2022-10-16 08:25:46,216] Trial 17 finished with value: 0.6607656115642149 and parameters: {'layer_size1': 512, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 128, 'learning_rate': 2.3798952412648256e-06, 'b1': 0.9980525150545131}. Best is trial 11 with value: 0.46340372127499985.


Train:  0.8135142980083343
Test 0.6607689261436462
0.6614920473926655
Epoch: 1596
Train:  0.8130209849726769
Test 0.6605861186981201
0.6613108616537564
Epoch: 1597
Train:  0.8125231804386261
Test 0.6604032516479492
0.661129339652595
Epoch: 1598
Train:  0.8120240492205466
Test 0.6602204442024231
0.6609475605625607
Epoch: 1599
Train:  0.8115128192209429
Test 0.6600378155708313
0.6607656115642149
Epoch: 0
Train:  0.6980514834004063
Test 0.6900963187217712
0.6900963187217712
Epoch: 1
Train:  1.0412802282840974
Test 0.6914088129997253
0.6908254822095237
Epoch: 2
Train:  1.040057219805256
Test 0.6927382946014404
0.6916094217144075
Epoch: 3
Train:  1.0389869713014173
Test 0.694089412689209
0.6924495270581749
Epoch: 4
Train:  1.038532558948763
Test 0.6956215500831604
0.6933931321941178
Epoch: 5
Train:  1.037752373564628
Test 0.6972915530204773
0.6944498209860662
Epoch: 6
Train:  1.0371971409167013
Test 0.6991074681282043
0.6956285472230224
Epoch: 7
Train:  1.0362956052826298
Test 0.70110911130

[I 2022-10-16 08:25:47,047] Trial 18 finished with value: 0.6900963187217712 and parameters: {'layer_size1': 384, 'layer_size2': 384, 'layer_size3': 256, 'layer_size4': 128, 'layer_size5': 64, 'learning_rate': 5.594360103803041e-05, 'b1': 0.9609196928489294}. Best is trial 11 with value: 0.46340372127499985.


Test 0.7234838008880615
0.712114223995109
Epoch: 15
Train:  1.0361745395968038
Test 0.7279590368270874
0.7153749683558033
Epoch: 16
Train:  1.0368868535564792
Test 0.7324461936950684
0.7188678664975033
Epoch: 17
Train:  1.038871628622855
Test 0.737126350402832
0.7225865531824436
Epoch: 0
Train:  0.6846435858357337
Test 0.7103950381278992
0.7103950381278991
Epoch: 1
Train:  1.0373644184681676
Test 0.717557966709137
0.7143744428952536
Epoch: 2
Train:  1.038367530030589
Test 0.7242490649223328
0.7184214191358599
Epoch: 3
Train:  1.0396320733331865
Test 0.7302464246749878
0.7224271798198463
Epoch: 4
Train:  1.0397402663384714
Test 0.7314172387123108
0.72510151909057
Epoch: 5
Train:  1.0376452563270446
Test 0.7328222990036011
0.7271942797140606
Epoch: 6
Train:  1.0353121719052714
Test 0.7362664341926575
0.7294901999895351
Epoch: 7
Train:  1.0333417798242261
Test 0.7333030700683594
0.7304065043705142
Epoch: 8
Train:  1.0267785768355093
Test 0.735419750213623
0.7315645890311131
Epoch: 9
Train

[I 2022-10-16 08:25:48,704] Trial 19 finished with value: 0.5618630652914107 and parameters: {'layer_size1': 512, 'layer_size2': 512, 'layer_size3': 384, 'layer_size4': 64, 'layer_size5': 128, 'learning_rate': 0.0001486254774652934, 'b1': 0.9212754626534484}. Best is trial 11 with value: 0.46340372127499985.


Train:  0.36950270413991904
Test 0.740239143371582
0.6356718988079805


In [ ]:
best_trial = study.best_trial

In [ ]:
print(best_trial.params)

{'layer_size1': 512, 'layer_size2': 512, 'layer_size3': 256, 'layer_size4': 64, 'layer_size5': 64, 'learning_rate': 4.531486810010387e-06, 'b1': 0.9591472429391925}


In [ ]:
import json

In [ ]:
with open('mydata.json', 'w') as f:
    json.dump(best_trial.params, f)

# Best Model training

In [ ]:
model = Net(768,[512,512,256,64,64],1).to(device)
optimizer = torch.optim.Adam(model.parameters(),lr=4.531486810010387e-06,betas=(0.9591472429391925,0.99))
lossff = torch.nn.BCELoss()

total_loss = 0
weighted_loss = 0
exp_param = 0.8

wloss = []
best_val = 1

for i in range(2500):
  print("Epoch:", i)
  model.train()
  for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data.x, data.edge_index, data.batch)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        loss.backward()
        optimizer.step()
        total_loss += float(loss) * data.num_graphs
  print("Train: ",total_loss / len(train_loader.dataset))

  model.eval()
  total_loss = 0
  for data in val_loader:
        data = data.to(device)
        out = model(data.x, data.edge_index, data.batch)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        total_loss += float(loss) * data.num_graphs
        
  print("Test", total_loss / len(val_loader.dataset))

  weighted_loss = exp_param*(weighted_loss) + (1-exp_param)*(total_loss/ len(val_loader.dataset))
  print(weighted_loss/(1-exp_param**(i+1)))
  wloss.append(weighted_loss/(1-exp_param**(i+1)))

  if(best_val>weighted_loss/(1-exp_param**(i+1))):
    best_val = weighted_loss/(1-exp_param**(i+1))

  if(best_val<0.50):
    break

Epoch: 0
Train:  0.6956658055705409
Test 0.6910093426704407
0.6910093426704407
Epoch: 1
Train:  1.0410125284425673
Test 0.6911178231239319
0.6910696095890471
Epoch: 2
Train:  1.0409150421619415
Test 0.6912251114845276
0.6911333398740801
Epoch: 3
Train:  1.0407743944275765
Test 0.6913238763809204
0.691197884761221
Epoch: 4
Train:  1.0406961921722657
Test 0.6914269924163818
0.6912660391089008
Epoch: 5
Train:  1.0405746679152212
Test 0.6915289163589478
0.6913372934593561
Epoch: 6
Train:  1.0404889564360342
Test 0.6916330456733704
0.6914121404547721
Epoch: 7
Train:  1.0403911384844011
Test 0.6917358636856079
0.6914899372347586
Epoch: 8
Train:  1.0403199964954006
Test 0.691839337348938
0.6915706503945169
Epoch: 9
Train:  1.0401819655972142
Test 0.6919368505477905
0.6916527004867399
Epoch: 10
Train:  1.0401239741233088
Test 0.6920366287231445
0.6917367017911238
Epoch: 11
Train:  1.0400205677555454
Test 0.6921361684799194
0.691822490482891
Epoch: 12
Train:  1.0399190668136842
Test 0.692236781

In [ ]:
from sklearn.metrics import accuracy_score, f1_score

In [ ]:
@torch.no_grad()
def val(model):
    model.eval()
    total_loss = 0
    all_preds = []
    all_labels = []
    for data in val_loader:
        data = data.to(device)
        out = model(data.x, data.edge_index, data.batch)
        # print(out)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        # print(loss)
        total_loss += float(loss) * data.num_graphs
        all_preds.append(torch.reshape(out, (-1,)))
        all_labels.append(data.y.float())
    # print(all_preds)
    accuracy, f1 = metrics(all_preds, all_labels)
    return total_loss / len(val_loader.dataset), accuracy, f1

@torch.no_grad()
def test(model):
    model.eval()
    total_loss = 0
    all_preds = []
    all_labels = []
    for data in test_loader:
        data = data.to(device)
        out = model(data.x, data.edge_index, data.batch)
        # print(out)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        # print(loss)
        total_loss += float(loss) * data.num_graphs
        all_preds.append(torch.reshape(out, (-1,)))
        all_labels.append(data.y.float())
    # print(all_preds)
    accuracy, f1 = metrics(all_preds, all_labels)
    return total_loss / len(test_loader.dataset), accuracy, f1


def metrics(preds, gts):
    preds = torch.round(torch.cat(preds))
    gts = torch.cat(gts)
    acc = accuracy_score(preds.cpu().numpy(), gts.cpu().numpy())
    f1 = f1_score(preds.cpu().numpy(), gts.cpu().numpy())
    return acc, f1

In [ ]:
val(model),test(model)

((0.4973466098308563, 0.8064516129032258, 0.8125000000000001),
 (0.47412919175570906, 0.8009049773755657, 0.7864077669902911))

In [ ]:
final = DataLoader(train_data_pol+val_data_pol, batch_size=32, shuffle=True)

In [ ]:
model = Net(768,[512,512,256,64,64],1).to(device)
optimizer = torch.optim.Adam(model.parameters(),lr=4.531486810010387e-06,betas=(0.9591472429391925,0.99))
lossff = torch.nn.BCELoss()

total_loss = 0
weighted_loss = 0
exp_param = 0.8

wloss = []

for i in range(1360):
  print("Epoch:", i)
  model.train()
  for data in final:
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data.x, data.edge_index, data.batch)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        loss.backward()
        optimizer.step()
        total_loss += float(loss) * data.num_graphs
  print("Train: ",total_loss / len(final.dataset))

  model.eval()
  total_loss = 0
  for data in test_loader:
        data = data.to(device)
        out = model(data.x, data.edge_index, data.batch)
        loss = lossff(torch.reshape(out,(-1,)), data.y.float())
        total_loss += float(loss) * data.num_graphs
        
  print("Test", total_loss / len(test_loader.dataset))

  weighted_loss = exp_param*(weighted_loss) + (1-exp_param)*(total_loss/ len(test_loader.dataset))
  print(weighted_loss/(1-exp_param**(i+1)))
  wloss.append(weighted_loss/(1-exp_param**(i+1)))

  # if(i-30>=0 and wloss[i-20]-weighted_loss<0.001):
  #   break

print(weighted_loss)

Epoch: 0
Train:  0.6932850473670549
Test 0.6930371671240794
0.6930371671240794
Epoch: 1
Train:  2.340089228845412
Test 0.6930178231243634
0.6930264204575707
Epoch: 2
Train:  2.3399687018445743
Test 0.6930029516306399
0.6930168020858777
Epoch: 3
Train:  2.3398405185309787
Test 0.692983571490551
0.6930055451091409
Epoch: 4
Train:  2.339719251278908
Test 0.6929662049625793
0.692993842304952
Epoch: 5
Train:  2.339601752578571
Test 0.6929479478710917
0.6929814023616947
Epoch: 6
Train:  2.3394957357837307
Test 0.6929307356139653
0.6929685799595103
Epoch: 7
Train:  2.33940339280713
Test 0.6929170195333558
0.6929561890195113
Epoch: 8
Train:  2.3393016822876467
Test 0.6929018354523775
0.6929436330758539
Epoch: 9
Train:  2.3391974741412747
Test 0.6928860453458933
0.6929307300813796
Epoch: 10
Train:  2.339099100840989
Test 0.6928698989600618
0.6929174205796205
Epoch: 11
Train:  2.339007043710319
Test 0.6928548327398516
0.6929039793364706
Epoch: 12
Train:  2.3389083050912425
Test 0.692839753304132

KeyboardInterrupt: ignored

In [ ]:
test(model)

(0.4395631611077494, 0.8552036199095022, 0.851851851851852)

In [ ]:
torch.save(model.state_dict(), "final.pt")

### Best Metrics found after training 
**Epoch: 114 |  TestLoss: 0.4395631611077494| TestAcc: 0.8552036199095022 | TestF1: 0.851851851851852**

*Note : We choose the best epoch based on lowest Validation Loss.*
